### ISSUE TITLE
sfn update fails with custom resources

### ISSUE THREAD

**Issue by ajohnstone** [sparkleformation/sfn#122]
```
$ sfn --version
sfn - SparkleFormation CLI - [Version: 2.1.0]

```
## Sample
```
SparkleFormation.dynamic(:ecs_custom_service) do |_name, _config = {}|

  resources("#{_name}_ecs_service".to_sym) do
    type 'Custom::EcsCreateService'
    creation_policy do
      resource_signal do
        timeout '30'
      end
    end
    update_policy do
      resource_signal do
        timeout '30'
      end
    end
    deletion_policy 'Delete'
    full_properties do
    end
    properties do
      service_token join!(
          "arn:aws:lambda:",
          region!,
          ":",
          account_id!,
          ":function:ecsCreateService",
      )
      cluster _config[:cluster].to_s if _config[:cluster]
      service do

        puts! "SERVICE:CLUSTER - #{_config[:service][:cluster].to_s}"

        cluster _config[:service][:cluster].to_s
        service_name _config[:service][:service_name]
        desired_count _config[:service][:desired_count]
        load_balancers _array(
          *_config[:service][:load_balancers].map { |lb| {
            'ContainerName' => lb[:container_name],
            'ContainerPort' => lb[:container_port],
            'LoadBalancerName' => lb[:load_balancer_name]
          }
        }) if (_config[:service][:load_balancers])
        role _config[:service][:role]
        task_definition _config[:service][:task_definition]
      end
    end
  end
end

```
## Exception

```
Custom::EcsCreateService

nil
"Cluster"
ERROR: NoMethodError: undefined method `[]' for nil:NilClass


/home/andrew/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/sfn-2.1.0/lib/sfn/planner/aws.rb:342:in `register_diff'
/home/andrew/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/sfn-2.1.0/lib/sfn/planner/aws.rb:304:in `block in run_stack_diff'
/home/andrew/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/sfn-2.1.0/lib/sfn/planner/aws.rb:303:in `each'
/home/andrew/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/sfn-2.1.0/lib/sfn/planner/aws.rb:303:in `run_stack_diff'
/home/andrew/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/sfn-2.1.0/lib/sfn/planner/aws.rb:192:in `plan_stack'
/home/andrew/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/sfn-2.1.0/lib/sfn/planner/aws.rb:278:in `block in run_stack_diff'
/home/andrew/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/sfn-2.1.0/lib/sfn/planner/aws.rb:262:in `each'
/home/andrew/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/sfn-2.1.0/lib/sfn/planner/aws.rb:262:in `run_stack_diff'
/home/andrew/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/sfn-2.1.0/lib/sfn/planner/aws.rb:192:in `plan_stack'
/home/andrew/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/sfn-2.1.0/lib/sfn/planner/aws.rb:134:in `generate_plan'
/home/andrew/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/sfn-2.1.0/lib/sfn/command/update.rb:94:in `execute!'
/home/andrew/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/sfn-2.1.0/bin/sfn:48:in `block (4 levels) in <top (required)>'
/home/andrew/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/bogo-cli-0.2.0/lib/bogo-cli/setup.rb:26:in `call'
/home/andrew/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/bogo-cli-0.2.0/lib/bogo-cli/setup.rb:26:in `block in bogo_cli_run'
/home/andrew/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/slop-3.6.0/lib/slop.rb:260:in `call'
/home/andrew/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/slop-3.6.0/lib/slop.rb:260:in `parse!'
/home/andrew/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/slop-3.6.0/lib/slop.rb:235:in `parse!'
/home/andrew/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/slop-3.6.0/lib/slop.rb:65:in `parse!'
/home/andrew/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/slop-3.6.0/lib/slop.rb:54:in `parse'
/home/andrew/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/bogo-cli-0.2.0/lib/bogo-cli/setup.rb:48:in `define'
/home/andrew/.rbenv/versions/2.2.4/lib/ruby/gems/2.2.0/gems/sfn-2.1.0/bin/sfn:14:in `<top (required)>'
/home/andrew/.rbenv/versions/2.2.4/bin/sfn:23:in `load'
/home/andrew/.rbenv/versions/2.2.4/bin/sfn:23:in `<main>'

```

**Comment by jimcroft** [comment]
Is it possible a key somewhere further down inside ```_config``` is missing? 

What does the whole config hash look like if you put  ```puts! "CONFIG: - #{_config.inspect}"``` in there?

**Comment by ajohnstone** [comment]
Sure... its

```
{
  :cluster=>"services-test-default",
  :service=> 
  {
    :cluster=>"services-test-default",
    :desired_count=>"1",
    :service_name=>"service-basket",
    :task_definition=>{"Fn::Join"=>["", ["arn:aws:ecs:", {"Ref"=>"AWS::Region"}, ":", {"Ref"=>"AWS::AccountId"}, ":task-definition/test-services-basket:latest"]]},
    :role=>{"Fn::Join"=>["", ["arn:aws:iam::", {"Ref"=>"AWS::AccountId"}, ":role/ecs-service-role"]]},
    :load_balancers=>[
      {:container_name=>"test-services-basket", :container_port=>"9080", :load_balancer_name=>{"Ref"=>"PublicElb"}}
    ]
  }
}
```

**Comment by chrisroberts** [comment]
Hi! You can side step the error by disabling the planner: `--no-plan`. There's two things I need to do here:

1. Rescue out plan generation exceptions and provide feedback on the failure but still allow the update
2. Proper type casting within the diff registration (since the error is on assumed type but `nil` being returned)

**Comment by chrisroberts** [comment]
Fixed in 2.1.2 release. Thanks for the report!
